modules/ml/src/main/java/org/apache/ignite/ml/tree/randomforest/RandomForestClassifierTrainer.java

```diff
@@ -17,72 +17,112 @@
-        return new DecisionTreeClassificationTrainer(maxDeep, minImpurityDecrease).withUseIndex(useIndex);
+    @Override protected void init(Dataset<EmptyContext, BootstrappedDatasetPartition> dataset) {
+        Set<Double> uniqLabels = dataset.compute(
+            x -> {
+                Set<Double> labels = new HashSet<>();
+                for (int i = 0; i < x.getRowsCount(); i++)
+                    labels.add(x.getRow(i).getLabel());
+                return labels;
+            },
+            (l, r) -> {
+                if (l == null)
+                    return r;
+                if (r == null)
+                    return l;
+                Set<Double> lbls = new HashSet<>();
+                lbls.addAll(l);
+                lbls.addAll(r);
+                return lbls;
+            }
+        );
+
+        int i = 0;
+        for (Double label : uniqLabels)
+            lblMapping.put(label, i++);
+
+        super.init(dataset);
     }
 
-    /**
-     * Sets useIndex parameter and returns trainer instance.
-     *
-     * @param useIndex Use index.
-     * @return Decision tree trainer.
-     */
-    public RandomForestClassifierTrainer withUseIndex(boolean useIndex) {
-        this.useIndex = useIndex;
-        return this;
+    /** {@inheritDoc} */
+    @Override protected ModelsComposition buildComposition(List<RandomForestTrainer.TreeRoot> models) {
+        return new ModelsComposition(models, new OnMajorityPredictionsAggregator());
+    }
+
+    /** {@inheritDoc} */
+    @Override protected GiniHistogram createImpurityComputer(int sampleId, BucketMeta meta) {
+        return new GiniHistogram(sampleId, lblMapping, meta);
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void addElementToLeafStatistic(ObjectHistogram<BootstrappedVector> leafStatAggr, BootstrappedVector vec, int sampleId) {
+        leafStatAggr.addElement(vec);
+    }
+
+    /** {@inheritDoc} */
+    @Override protected ObjectHistogram<BootstrappedVector> mergeLeafStats(ObjectHistogram<BootstrappedVector> leafStatAggr1,
+        ObjectHistogram<BootstrappedVector> leafStatAggr2) {
+
+        leafStatAggr1.addHist(leafStatAggr2);
+        return leafStatAggr1;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected ObjectHistogram<BootstrappedVector> createLeafStatsAggregator(int sampleId) {
+        return new ObjectHistogram<>(
+            x -> lblMapping.get(x.getLabel()),
+            x -> (double)x.getRepetitionsCounters()[sampleId]
+        );
+    }
+
+    /** {@inheritDoc} */
+    @Override protected double computeLeafValue(ObjectHistogram<BootstrappedVector> stat) {
+        Integer bucketId = stat.buckets().stream()
+            .max(Comparator.comparing(b -> stat.get(b).orElse(0.0)))
+            .get();
+
+        return lblMapping.entrySet().stream()
+            .filter(x -> x.getValue().equals(bucketId))
+            .findFirst()
+            .get().getKey();
```

Comment:
Is it possible mapping to be empty here?

Reply:
no, this mapping built from dataset and histogram uses same mapping